x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -127,12 +209,354 @@ func (k msgServer) queryAndValidateClaimForProof(ctx context.Context, proof *typ
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay request service ID %s does not match proof service ID %s",
+			reqSessHeader.GetService().GetId(),
+			msgSessHeader.GetService().GetId(),
+		)
+	}
+
+	if respSessHeader.GetService().GetId() != msgSessHeader.GetService().GetId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay response service ID %s does not match proof service ID %s",
+			respSessHeader.GetService().GetId(),
+			msgSessHeader.GetService().GetId(),
+		)
+	}
+
+	// Ensure the relay request and response session start block heights
+	// match the proof session start block height.
+
+	if reqSessHeader.GetSessionStartBlockHeight() != msgSessHeader.GetSessionStartBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay request session start height %d does not match proof session start height %d",
+			reqSessHeader.GetSessionStartBlockHeight(),
+			msgSessHeader.GetSessionStartBlockHeight(),
+		)
+	}
+
+	if respSessHeader.GetSessionStartBlockHeight() != msgSessHeader.GetSessionStartBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay response session start height %d does not match proof session start height %d",
+			respSessHeader.GetSessionStartBlockHeight(),
+			msgSessHeader.GetSessionStartBlockHeight(),
+		)
+	}
+
+	// Ensure the relay request and response session end block heights
+	// match the proof session end block height.
+
+	if reqSessHeader.GetSessionEndBlockHeight() != msgSessHeader.GetSessionEndBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay request session end height %d does not match proof session end height %d",
+			reqSessHeader.GetSessionEndBlockHeight(),
+			msgSessHeader.GetSessionEndBlockHeight(),
+		)
+	}
+
+	if respSessHeader.GetSessionEndBlockHeight() != msgSessHeader.GetSessionEndBlockHeight() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay response session end height %d does not match proof session end height %d",
+			respSessHeader.GetSessionEndBlockHeight(),
+			msgSessHeader.GetSessionEndBlockHeight(),
+		)
+	}
+
+	// Ensure the relay request and response session IDs match the proof session ID.
+
+	if reqSessHeader.GetSessionId() != msgSessHeader.GetSessionId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay request session ID %s does not match proof session ID %s",
+			reqSessHeader.GetSessionId(),
+			msgSessHeader.GetSessionId(),
+		)
+	}
+
+	if respSessHeader.GetSessionId() != msgSessHeader.GetSessionId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay response session ID %s does not match proof session ID %s",
+			respSessHeader.GetSessionId(),
+			msgSessHeader.GetSessionId(),
+		)
+	}
+
+	return nil
+}
+
+// verifyRelayRequestSignature verifies the signature on the relay request.
+// TODO_TECHDEBT: Factor out the relay request signature verification into a shared
+// function that can be used by both the proof and relayer packages.
+func (k msgServer) verifyRelayRequestSignature(
```

Comment:
As discussed, after #411, we should be able to move this to a method on the relay request type for use both on- and off-chain.